src/sidebar/cross-origin-rpc.js

```diff
@@ -0,0 +1,70 @@
+'use strict';
+
+/**
+ * Begin responding to JSON-RPC requests from frames on other origins.
+ *
+ * Register a window.postMessage() event listener that receives and responds to
+ * JSON-RPC requests sent by frames on other origins using postMessage() as the
+ * transport layer.
+ *
+ * Only frames whose origin is in the rpcAllowedOrigins config setting will be
+ * responded to.
+ *
+ * This is a very partial implementation of a JSON-RPC 2.0 server:
+ *
+ * http://www.jsonrpc.org/specification
+ *
+ * The only part that we support so far is receiving JSON-RPC 2.0 requests (not
+ * notifications) without any parameters and sending back a successful
+ * response. Notifications (JSON-RPC calls that don't require a response),
+ * method parameters, and error responses are not yet supported.
+ *
+ */
+// @ngInject
+function start(annotationUI, settings, $window) {
+  $window.addEventListener('message', function receiveMessage(event) {
+    let allowedOrigins = settings.rpcAllowedOrigins || [];
+
+    if (!allowedOrigins.includes(event.origin)) {
+      return;
+    }
+
+    // The entire JSON-RPC request object is contained in the postMessage()
+    // data param.
+    let jsonRpcRequest = event.data;
+
+    event.source.postMessage(jsonRpcResponse(jsonRpcRequest), event.origin);
+  });
```

Comment:
Things can get a bit confusing because there are actually two separate protocols being used together here:

1. [`window.postMessage()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) is used to send messages between frames on different origins. All the `postMessage()` considerations are kept together in this `receiveMessage()` function, In particular we check that the `origin` of the message we've received is in the `rpcAllowedOrigins` config setting, and ignore the message if not. We also use `event.source.postMessage()` to send our response back to the caller in a way that only code on the caller's origin (`event.origin`) will be able to receive the response.

2. [JSON-RPC 2.0](http://www.jsonrpc.org/specification). The `data` param of the `window.postMessage()` message is a JSON-RPC request object. We pass this RPC request object into our `jsonRpcResponse()` function below. All of the RPC-related considerations are kept together in the `jsonRpcResponse()` function. In particular it only responds to `"searchUris"` requests, and sends back an error message otherwise. It also preserves the request's `id` in the response, so that when the sender receives the response it knows which request it is in response to.

Reply:
This is a really nice idea.